minetest/minetest
# Animated particlespawners

fixes #11505, adding many new features to particlespawners and cleaning up some of the particle code while maintaining backward-compatibility with older servers and clients. with this PR, minetest particles take a step towards parity with the particle systems of other game engines, giving much more control and flexibility to mod and game devs, particularly with the new animation system, which could be extended to a full keyframe system without too much difficulty. i believe this is all in line with the "graphics improvement" part of the roadmap?

see the lua_api.txt diff for a full list of new particlespawner parameters

## To do

this PR is a ready for review i think

## caveats

i've never contributed to the minetest engine before so i had a limited understanding of where things were supposed to go, and i may have duplicated mechanisms that already existed within the codebase simply because i couldn't find them. this is also the first thing i've written in C++ in quite a few years. so there are probably a lot of organizational issues that will need addressing and i may have written parts in a insufficiently idiomatic way, sorry.

## How to test

this PR adds a very wide range of features. the ideal way to test it would be to take an existing mod that uses particles, and update its `minetest.add_particlespawner` calls to use the new parameters. the `doc/lua_api.txt` file thoroughly documents all the new additions. testing with patched clients against older servers and with patched servers against older clients to make sure particlespawners continue to behave as expected is also important.
